### PR TITLE
fix: whitelisting and adding observedAttributes in render

### LINKED ISSUE #54: Current React lifecycle methods problem
I've been trying to "fix a bug" when using `react-integration` with `skatejs` components. 
I found what is the problem but I'm not sure how to solve it properly. If anyone have any idea / feedback on this I will appreciate it. 

You can see the issue [here](http://codepen.io/mcorbanini/pen/xEdWdR?editors=1011). 

So, if you take a look at the console, that button is rendered once without / with missing values and a second time with all the needed properties. 

That MAY cause some weird behaviour (in my case, my component flashes when it is instanciated with attributes that changes a default style applied previously). 

If you don't use react-integration it won't happen. So digging a bit inside `react-integration` I found that lifecycle methods in react don't work in sync with skate's or at least that is my sensation. 

Trying to make this message not too long, the sequence I see is:

1. [Render](https://github.com/webcomponents/react-integration/blob/master/src/index.js#L70)
Initial render, that create a react component with only `style` prop. This is to avoid dynamically computed properties using `get` and `set`
-> 
2. [componentDidMount](https://github.com/webcomponents/react-integration/blob/master/src/index.js#L52) 
-> 
3. [componentWillReceiveProps](https://github.com/webcomponents/react-integration/blob/master/src/index.js#L55) (called explicitly in `componentDidMount`) This callback add all the attributes to the current `node`, where `node` is the result of `ReactDOM.findDOMNode(this)` ( see [here](https://facebook.github.io/react/docs/top-level-api.html#reactdom.finddomnode) ). That will trigger another render in skatejs, but not in the tests (look at the props tests if you are interested) because tests are not using skatejs components only native custom elements. 

Naturally I will think of this sequence to be:
`componentWillMount` (call `componentWillReceiveProps`) 
->
`componentWillReceiveProps` Here it should set all props in the `node`
->
`render` at this point everything should be in place

The problem is, you cannot access `findDOMNode` in `componentWillReceiveProps` to add all the properties, so this approach I guess is not possible.

Will be looking forward to any ideas / feedback / correction on any wrong information that I may have written.
Thanks!

### PR DESCRIPTION
**Pull request by mcorbanini**
fixes #54

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -27,6 +27,13 @@ function syncEvent(node, eventName, newEventHandler) {
 }
 
 export default function (CustomElement, opts) {
+  function whitelistAttrs(props) {
```
Comment (treshugart) [inline src/index.js]: I'm unsure if this is the best name for it, but I don't want to 🚲 shed too much on it.

File: src/index.js [hunk 2]
```diff
@@ -43,25 +50,34 @@ export default function (CustomElement, opts) {
...
         if (name.indexOf('on') === 0 && name[2] === name[2].toUpperCase()) {
           syncEvent(node, name.substring(2), props[name]);
         } else {
           node[name] = props[name];
         }
       });
     }
+    onRef(node) {
+      if (node != null) {
+          this.applyProps(node, this.props);
```
Comment (treshugart) [inline src/index.js]: nit: indentation (eslint should have picked this up, though)

File: src/index.js [hunk 3]
```diff
@@ -43,25 +50,34 @@ export default function (CustomElement, opts) {
...
-
         if (name.indexOf('on') === 0 && name[2] === name[2].toUpperCase()) {
           syncEvent(node, name.substring(2), props[name]);
         } else {
           node[name] = props[name];
         }
       });
     }
+    onRef(node) {
+      if (node != null) {
```
Comment (treshugart) [inline src/index.js]: Can node ever be anything but an element or falsy?

File: src/index.js [hunk 4]
```diff
@@ -43,25 +50,34 @@ export default function (CustomElement, opts) {
     static get displayName() {
       return displayName;
     }
-    componentDidMount() {
-      this.componentWillReceiveProps(this.props);
+    constructor(props) {
+      super(props);
```
Comment (treshugart) [inline src/index.js]: This seems unrelated to this particular fix?

File: src/index.js (hunk 3 above)
Comment (mcorbanini) [inline src/index.js]: I guess this is just being defensive. `ref` callback will always have the element right?

Comment (treshugart) [inline src/index.js]: Oh wait, I was conflating this with Skate. React does call ref with null (but I think it's always `null`). You could probably just use `if (node)`. If you want to be 100% safe, `if (node instanceof HTMLElement)` would work, but I think that's overkill and truthy / falsy checks here work just fine.

**Comment by treshugart** [comment]
Closing as I think this doesn't solve a concrete purpose. Happy to revisit if I'm wrong.
